bindings/pyroot/ROOT.py

```diff
@@ -265,39 +265,85 @@ def _proxy__array_interface__(self):
         class_scope.__array_interface__ = property(class_scope._proxy__array_interface__)
 
 # TTree.AsMatrix functionality
-def _TTreeAsMatrix(self, columns):
+def _TTreeAsMatrix(self, columns=None):
+    # Import numpy lazily
     try:
         import numpy as np
     except:
         raise ImportError("Failed to import numpy during call of TTree.AsMatrix.")
 
-    # Error-handling
+    # Check that tree has entries
     if self.GetEntries() == 0:
         raise Exception("Tree has no entries.")
 
+    # Get all columns of the tree if no columns are specified
+    if columns is None:
+        columns = [branch.GetName() for branch in self.GetListOfBranches()]
+
+    # Check that all given columns exist
+    for col in columns:
+        branch = self.GetBranch(col)
+        if branch == None:
+            raise Exception("Tree {} has no branch {}.".format(self.GetName(), col))
+
+    # Check that all branches have allowed types and find out whether there are
+    # only signed or unsigned types
+    allowed_dtypes_unsigned = [
+            "UInt_t", "unsigned int",
+            "ULong64_t", "unsigned long"]
+    allowed_dtypes_signed = [
+            "Int_t", "int",
+            "Long64_t", "long",
+            "Float_t", "float",
+            "Double_t", "double"]
+    allowed_dtypes = allowed_dtypes_unsigned + allowed_dtypes_signed
+    col_dtypes = []
+    has_signed_dtype = False
+    for i, col in enumerate(columns):
+        name_dtype = self.GetLeaf(col).GetTypeName()
+        if not name_dtype in allowed_dtypes:
+            raise Exception("Branch {} of tree {} has not supported data-type {}.".format(
+                    col, self.GetName(), name_dtype))
+        if name_dtype in allowed_dtypes_signed:
+            has_signed_dtype = True
+        col_dtypes.append(name_dtype)
+
     # Convert columns interable to std.vector("string")
+    # and find correct data-type for output array
     columns_vector = _root.std.vector("string")(len(columns))
     for i, col in enumerate(columns):
         columns_vector[i] = col
 
-    # Allocate memory for the read-out
+    # Allocate memory for the read-out with proper data-type
     # NOTE: This has to be done in PyROOT for a proper handling of the reference counting
-    dtype = "double" # TODO: infere the correct data-type for the output array
-    flat_matrix = _root.std.vector(dtype)(self.GetEntries()*len(columns))
-
-    # Read the tree as flat std.vector(dtype)
+    usable_dtypes = allowed_dtypes if has_signed_dtype else allowed_dtypes_unsigned
+    output_dtype = usable_dtypes[max([usable_dtypes.index(dtype) for dtype in col_dtypes])]
+    root_dtype_conversion = {
```

Comment:
Why do we need this last type conversion?

Reply:
First, we need a ROOT type conversion (`Float_t`) to `std.vector` types (`float`). Then, we need to ensure that we return a signed type, if one of the columns is signed. There can be corner-cases, e.g., the columns with types `int, unsigned long` to output type `long`.

Reply:
I prefer the sign over the precision of the data-type, which seems feasible to me.

Reply:
However, implementing this type-conversion is kind of weird because there's not a "right way" to do it at the end (?)

Reply:
If we end up encapsulating all this inside a multi-dimensional `Take` on TDF that returns a `MultiVec`, perhaps it is not worth the trouble of going through this type conversion here? If what we wanted to support is getting ntuples as numpy arrays, why not just supporting columns of type float/double?

Reply:
we can convert anything by default to `np.float64`, that's correct. that would make everything much easier. to be honest, i think that most users won't notice (and don't care about) the type conversion stuff... still, from the technical side, it's nice to have! and with a look on the tests, the implementation seems to work nicely.

Reply:
Option 1: Use this kind of complicated type-conversion.
Option 2: Hard-code the return type as `np.float64` (default data-type for all numpy arrays!)